Utilities::MPI::mpi_sum_avg: create MPI_Datatype and MPI_Op once

tamiko opened the pull request:
Instead of repeatedly creating and freeing an MPI_Datatype and MPI_Op for
every invocation of Utilities::MPI::mpi_sum_avg let us define these objects
statically and create them only once.

This works around a memory leak issue with OpenMPI 4.1.0 that will leak 
memory in the subsequent MPI_Allreduce call.

Upstream bug report: https://github.com/open-mpi/ompi/issues/8827

tamiko commented:
/rebuild

tamiko commented:
@tjhei *ping*

tjhei reviewed (commented):
I am okay with merging, but consider my two suggestions.

tamiko commented on source/base/mpi.cc:
```diff
@@ -605,6 +605,45 @@ namespace Utilities
+       * of the MinMaxAvg struct and a custom MPI op handle for our
+       * max_reduce function.
+       *
+       * This setup will leak some minor allocated memory (in the range of
```
What about we refactor all of this into its own `static struct` in the cc file?

tamiko replied:
Actually. I like this approach. Let me refactor.

tamiko replied:
@tjhei I know that this is not elegant. The entire MPI context however is already leaking like crazy.

bangerth replied:
You can always to a `static std::unique_ptr<...>` local variable. That is initialized on first run-through of the function, and deallocated at the end of the program.

tamiko replied:
@bangerth I had that variant before and reverted to this first version. The problem with this is that the deallocation happens after `dealii::MPI_InitFinalize` closed the MPI context and the program will abort with an error.

masterleinad replied:
I would also prefer to use a static variable in `MPI_InitFinalize` (or a callback) that can handle the cleanup in the destructor of `MPI_InitFinalize`.

bangerth replied:
I see. What we need is a `static` member variable of type `boost::signal` in the `MPI_InitFinalize` class, say called `at_mpi_exit` that one can attach to. That's where you would put the call that de-allocates the memory.

tamiko replied:
Implementing.

tamiko commented on source/base/mpi.cc:
```diff
@@ -605,6 +605,45 @@ namespace Utilities
+          ierr = MPI_Type_commit(&type);
+          AssertThrowMPI(ierr);
+        });
+
```
Refactored.

bangerth commented on source/base/mpi.cc:
```diff
@@ -622,13 +622,49 @@ namespace Utilities
+       * complain about this (such as clang's -fsanitize=address).
+       */
+      static MPI_Datatype type;
+      static MPI_Op       op;
```
You can do this easier than using `std::call_once` if you do 
```suggestion
      static MPI_Op       op = [=]() {...} ();
```
In other words, you just provide an initializer for the static variable.

tamiko replied:
I thought about this as well.

I opted for the `std::call_once` variant because I think it is mildly cleaner. (We cannot call to `MPI_Type_create_struct`, `MPI_Type_commit` and `MPI_Op_create` outside of an `MPI_InitFinalize` so all of the static initializer would need a separate `if(job_supports_mpi())`.)

bangerth replied:
No, that's what local `static` variables do: They are initialized upon first go-through of the function. They do exactly what you are doing with the `call_once`.

tamiko replied:
I will revise.

tamiko replied:
`std::call_once` is now gone.

tjhei replied:
> They are initialized upon first go-through of the function.

Yes, and the compiler even generates code to make it thread-safe.

bangerth commented on source/base/mpi.cc:
```diff
@@ -622,13 +622,49 @@ namespace Utilities
+      static MPI_Datatype type;
+      static MPI_Op       op;
+
+      std::once_flag initialization_flag;
```
This isn't going to work unless you mark the variable as `static`. But I think there is a better option (see above).

bangerth reviewed (approved):
Looks good to me. We *could* deal with the memory leak, but I'm not actually too concerned about it. My major worry is that having these kinds of leaks makes it more difficult to run tools such as `memcheck` on deal.II-based codes. But let's do this one step at a time.

tamiko commented:
@bangerth I am now halfway through implementing the signal handlers. Let me add that to the pull request. :smile:

tamiko commented:
@bangerth @masterleinad I will finish up the singal handler this evening. (Tests and hooking up the deallocator are missing).

kronbichler commented on doc/news/changes/minor/20210419Maier:
```diff
@@ -0,0 +1,8 @@
+Added: The MPI_InitFinalize RAII class has gained a
+MPI_InitFinalize::singals::at_mpi_ini and an
```
```suggestion
MPI_InitFinalize::singals::at_mpi_init and an
```

jppelteret replied:
singals -> signals

masterleinad commented on doc/news/changes/minor/20210419Maier:
```diff
@@ -0,0 +1,8 @@
+Added: The MPI_InitFinalize RAII class has gained a
+MPI_InitFinalize::singnals::at_mpi_init and an
```
```suggestion
MPI_InitFinalize::signals::at_mpi_init and an
```

tamiko replied:
Fixed!

masterleinad reviewed (commented):
Apart from the typo, this looks good to me.

tamiko commented:
@dealii/dealii Alright. I have added a small test for the signal handler and I think I have addressed all typos. Let's see whether the testsuite likes it.

masterleinad commented on doc/news/changes/minor/20210419Maier:
```diff
@@ -0,0 +1,8 @@
+Added: The MPI_InitFinalize RAII class has gained an
+MPI_InitFinalize::signals::at_mpi_init and an
+MPI_InitFinalize::signals::at_mpi_finalize signal. That are triggered
```
```suggestion
MPI_InitFinalize::signals::at_mpi_finalize signal that are triggered
```
nitpick: not a sentence. :slightly_smiling_face:

masterleinad commented on tests/base/mpi_init_finalize_signals_01.cc:
```diff
@@ -0,0 +1,42 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2017 - 2018 by the deal.II authors
```
```suggestion
// Copyright (C) 2021 by the deal.II authors
```

bangerth commented:
Thanks, @tamiko , for sticking with everyone's escalating requests!
